### PR TITLE
SOLR-14940: Add ability to remove SolrCore.closeHooks. Keep reference…

### PR DESCRIPTION
**Pull request by anverus**
…s to CloseHooks in ReplicationHandler and remove them on ReplicationHandler.shutdown()

# Description

Please provide a short description of the changes you're making with this pull request.

# Solution

Please provide a short description of the approach taken to implement your solution.

# Tests

Please describe the tests you've developed or run to confirm this patch implements the feature or solves the problem.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `master` branch.
- [x] I have run `./gradlew check`.
- [ ] I have added tests for my changes.
- [ ] I have added documentation for the [Ref Guide](https://github.com/apache/lucene-solr/tree/master/solr/solr-ref-guide) (for Solr changes only).

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/handler/ReplicationHandler.java [hunk 1]
```diff
@@ -1372,55 +1372,55 @@ private boolean isEnabled( @SuppressWarnings({"rawtypes"})NamedList params ){
...
-    }
-    ExecutorUtil.shutdownAndAwaitTermination(restoreExecutor);
-    if (restoreFuture != null) {
-      restoreFuture.cancel(false);
-    }
-    
+    startShutdownHook.preClose(core);
+    startShutdownHook.postClose(core);
+    finishShutdownHook.preClose(core);
+    finishShutdownHook.postClose(core);
```
Comment (anverus) [inline solr/core/src/java/org/apache/solr/handler/ReplicationHandler.java:1418]: Order of these calls copies order of execution of instructions in original shutdown method, while the correct order seems to be 
`
    startShutdownHook.preClose(core);
    finishShutdownHook.preClose(core);
    startShutdownHook.postClose(core);
    finishShutdownHook.postClose(core);
`

File: solr/core/src/java/org/apache/solr/handler/ReplicationHandler.java [hunk 2]
```diff
@@ -415,19 +415,15 @@ private void fetchIndex(SolrParams solrParams, SolrQueryResponse rsp) throws Int
...
     checksum.reset();
     byte[] buffer = new byte[1024 * 1024];
     int bytesRead;
-    try {
-      fis = new FileInputStream(f);
+    try (final FileInputStream fis = new FileInputStream(f)) {
       while ((bytesRead = fis.read(buffer)) >= 0)
         checksum.update(buffer, 0, bytesRead);
       return checksum.getValue();
     } catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (madrob) [reply solr/core/src/java/org/apache/solr/handler/ReplicationHandler.java]: @TomMD I think this is a bug in Muse?

[the remark this answers is not in the document]
Comment (TomMD) [reply solr/core/src/java/org/apache/solr/handler/ReplicationHandler.java]: @madrob Could you say more about why you think it's a bug? Do you think it is a false positive?  I don't have nearly the same view into the code as you do, so perhaps you are right but my unfamiliar reading leads me to another conclusion.

This Muse result originates from Infer - I can get the full trace if you'd like but by eye I'd expect it to say:

1. `fis` is allocated a new FileInputStream on line 421.
2. The fall to `fis.read()` could throw an IO exception.
3. There are no further references to `fis` in the code path.

Notably, the prior code of `finally { IOUtils.closeQuietly(fis); }` has been removed.  S not only Infer, but also some prior author, believe `fis` needs closed and now a path exists where `fis` is not closed.

[the remark this answers is not in the document]
Comment (dweiss) [reply solr/core/src/java/org/apache/solr/handler/ReplicationHandler.java]: This handle is always released in the finally block. I don't think an automated tool can "see" this from this code alone but the IOUtils.closeQuietly(fis); in finally ensures fis is always closed. The replacement try-with-resources isn't exactly the same as previous version would omit exceptions during close while the replacement would rethrow an exception during close attempt.

[the remark this answers is not in the document]
Comment (TomMD) [reply solr/core/src/java/org/apache/solr/handler/ReplicationHandler.java]: @dweiss Are we talking about the same code?  The pull request / commit Muse commented on actually deleted the finally block and its `closeQuietly` call.  [Here is the code analyzed](https://github.com/apache/lucene-solr/blob/c244474b2d3934a4b3d1ee2fbb419cf90605495d/solr/core/src/java/org/apache/solr/handler/ReplicationHandler.java#L421) which was flagged with the bug.

That change was subsequently reverted by @anver-trip (commit 8a10d1b) so this PR no longer has the bug in the comment.  Github helpfully adds an "Outdated" tag to comments that are on lines which have changed, such as this one.

Please let me know if I've misunderstood the situation or we can make things flow more smoothly!

[the remark this answers is not in the document]
Comment (dweiss) [reply solr/core/src/java/org/apache/solr/handler/ReplicationHandler.java]: Ok, I'm confused then. I was looking at the closeQuietly version... these are not the same in what they do. Please disregard.

[the remark this answers is not in the document]
Comment (madrob) [reply solr/core/src/java/org/apache/solr/handler/ReplicationHandler.java]: > There are no further references to fis in the code path.

The try-with-resources will generate a synthetic finally block that calls `fis.close()`. If the call to close throws an Exception, then... I guess we didn't release it? But in the closeQuietly case, we swallow the exception and we don't guarantee the close either.

[the remark this answers is not in the document]
Comment (TomMD) [reply solr/core/src/java/org/apache/solr/handler/ReplicationHandler.java]: @madrob Oh I see.  It seems quite likely that Infer doesn't understand try-with-resources (and yes, I missed it in my glance at the code).  So this is a false positive for that reason.  Thank you, I'll run with this example to try and get things cleaned up.

In general there does exist some level of false positives.  We're working hard to get and keep those quite minimal.  Everyone is encouraged to use the 'resolve' button and we are simultaneously working on a command system so you can state "ignore" to get to bot to resolve and adjust the analysis.  Thanks again for noting this instance, we will use it to learn.

**Pull request closed without merging**